### ISSUE TITLE
Deadlock if commit is called inside a MapListener

### ISSUE THREAD

**Issue by atoulme** [jankotek/mapdb#607]
I have a case where a record is committed, upon which a MapListener triggers, and runs code which ends up modifying the same cache. I now need to commit the data again. All this interaction should take place in the same thread to avoid race conditions and make it possible to run transaction replays.

With that in mind, it looks like the MapListener should have ways to run either before commit acquires locks, or after commit has released locks. I'm more interested in the latter case.

Here is the test case I use to reproduce, I'll take a stab at figuring out a solution.

```java
    @Test
    public void testListenerDeadlock() {
        final DB db = DBMaker.memoryDB().make();
        final HTreeMap map = db.hashMap("test");
        map.modificationListenerAdd(new MapListener() {
            @Override
            public void update(Object key, Object oldVal, Object newVal) {
                if ("foo".equals(newVal)) {
                    map.put("xyz", "bar");
                }
                db.commit();
            }
        });
        map.put("abc", "foo");
    }

**Comment by atoulme** [comment]
I narrowed down the issue to the following. HTreeMap is built with a consistency lock which is the read consistency lock of the DB. When db.commit() is called, DB attempts to lock a write lock, which waits for the read lock to be disposed. Hence the deadlock right now.

I don't have a good grasp on the solution. Passing the write lock to the HTreeMap instead of the read lock does the job (the lock is not used for get, only for put, replace, remove... so it's effectively writing to the map, correct?)

However, I guess the write lock was meant to be used to store the DB back to disk.

I'll open a PR to show the changes. As far as I can tell, all non-failing tests still pass.

**Comment by shabanovd** [comment]
Wander, why you have two commit calls?

**Comment by atoulme** [comment]
@shabanovd the second call was never reached. I updated the code to make the problem explicit.

@jankotek thanks for reviewing the PR. I have pushed an other commit where I take the view of moving the calls to notify outside the read locks - it looks like this is an efficient way to fix the issue. Is there a reason why listeners should be notified while the lock is on? See 25451672774617a6f458c23add3f693b5a0254e6
